src/software/amazon/ionschema/internal/util/RangeBigDecimal.kt

```diff
@@ -0,0 +1,120 @@
+
+import software.amazon.ion.*
+import software.amazon.ionschema.InvalidSchemaException
+import java.math.BigDecimal
+
+/**
+ * Implementation of Range<BigDecimal>.  As of this writing, all other Range<T>
+ * implementations directly or indirectly delegate to an instance of this class.
+ */
+internal class RangeBigDecimal(private val ion: IonList) : Range<BigDecimal> {
+    companion object {
+        private fun toBigDecimal(ion: IonValue) =
+                if (ion.isNullValue) {
+                    throw InvalidSchemaException("Unable to convert $ion to BigDecimal")
+                } else {
+                    when (ion) {
+                        is IonDecimal -> ion.bigDecimalValue()
+                        is IonFloat -> ion.bigDecimalValue()
+                        is IonInt -> BigDecimal(ion.bigIntegerValue())
+                        else ->
+                            throw InvalidSchemaException(
+                                    "Expected range lower/upper to be a decimal, float, or int (was $ion)")
+                    }
+                }
+    }
+
+    internal val lower: Boundary
+    internal val upper: Boundary
+    init {
+        if (!ion.hasTypeAnnotation("range")) {
+            throw InvalidSchemaException("Invalid range, missing 'range' annotation:  $ion")
+        }
+        if (ion.size != 2) {
+            throw InvalidSchemaException("Invalid range, size of list must be 2:  $ion")
+        }
+
+        lower = if (ion[0] is IonSymbol && (ion[0] as IonSymbol).stringValue() == "min") {
+                Boundary(null, Infinity.NEGATIVE)
+            } else {
+                Boundary(ion[0], Infinity.NEGATIVE)
+            }
+        upper = if (ion[1] is IonSymbol && (ion[1] as IonSymbol).stringValue() == "max") {
+                Boundary(null, Infinity.POSITIVE)
+            } else {
+                Boundary(ion[1], Infinity.POSITIVE)
+            }
+
+
+        if (lower > upper) {
+            throw InvalidSchemaException("Lower bound must be <= upper in $ion")
+        }
+        if (lower.value != null && upper.value != null
+                && lower.value == upper.value
+                && (lower.boundaryType == RangeBoundaryType.EXCLUSIVE
+                    || upper.boundaryType == RangeBoundaryType.EXCLUSIVE)) {
+            throw InvalidSchemaException("No valid values in $ion")
+        }
+    }
+
+    override fun contains(value: BigDecimal) = lower <= value && upper >= value
+
+    override fun toString() = ion.toString()
+
+    internal enum class Infinity(val sign: Int) {
+        NEGATIVE(-1),
+        POSITIVE( 1),
+    }
+
+    /**
+     * Represents either a lower or upper boundary of a range.
+     */
+    internal class Boundary(ion: IonValue?, private val infinity: Infinity) : Comparable<Boundary> {
+        internal val value: BigDecimal?
+        internal val boundaryType: RangeBoundaryType
+        init {
+            value = if (ion == null || ion.isNullValue) {
```

Comment:
I see this pattern being repeated a lot.  You can  add an extension property:

```
val IonValue?.isNullOrIonNull get() this == null || this.isNullValue
```
Or you can make it a function if you prefer:
```
fun IonValue?.isNullOrIonNull() = this == null || this.isNullValue